File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.Arrays;

class Solution {
    static final int INF = 20000001; // 충분히 큰 값 (최대 비용 100000, n=200이므로)

    public int solution(int n, int s, int a, int b, int[][] fares) {
        int[][] dist = new int[n + 1][n + 1];

        // 초기화
        for (int i = 1; i <= n; i++) {
            Arrays.fill(dist[i], INF);
            dist[i][i] = 0;
        }

        // 간선 정보 입력
        for (int[] fare : fares) {
            int c = fare[0];
            int d = fare[1];
            int f = fare[2];
            dist[c][d] = f;
            dist[d][c] = f;
        }

        // 플로이드-워셜
        for (int k = 1; k <= n; k++) {
            for (int i = 1; i <= n; i++) {
                for (int j = 1; j <= n; j++) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }

        // 최솟값 탐색
        int answer = INF;
        for (int i = 1; i <= n; i++) {
            answer = Math.min(answer, dist[s][i] + dist[i][a] + dist[i][b]);
        }

        return answer;
    }
}

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
def solution(n, s, a, b, fares):
    #무한 정의
    INF = int(1e9)
    #graph 만들고
    graph = [[INF] * (n+1) for _ in range(n+1)]
    #graph 초기화
    for i in range(1, n+1):
        graph[i][i]=0
    #graph에 fares 값 넣고
    for c,d,f in fares:
        graph[c][d] = f
        graph[d][c] = f
    #3중 for문 min 
    for k in range(1, n+1):
        for i in range(1, n+1):
            for j in range(1, n+1):
                graph[i][j]=min(graph[i][j], graph[i][k]+graph[k][j])
    #answer min 값 구하기
    answer = INF
    for i in range(1, n+1):
        answer=min(answer, graph[s][i]+graph[i][a]+ graph[i][b])
    return answer